### PR TITLE
Vad prob states

### PR DESCRIPTION
**Pull request by sarane22**
Exposing VAD states/probabilities to client.

### CODE REVIEW

File: riva/clients/asr/client_call.cc [hunk 1]
```diff
@@ -13,45 +13,63 @@ ClientCall::ClientCall(uint32_t corr_id, bool word_time_offsets)
   recv_final_flags.reserve(1000);
 }
 
+ClientCall::~ClientCall(){
```
Comment (virajkarandikar) [inline riva/clients/asr/client_call.cc]:
```suggestion
ClientCall::~ClientCall() {
```

File: riva/clients/asr/client_call.cc [hunk 2]
```diff
@@ -13,45 +13,63 @@ ClientCall::ClientCall(uint32_t corr_id, bool word_time_offsets)
...
+    std::string vad_log = "";
+    for (int i = 0; i < prob_states_count; i++) {
+      vad_log += std::to_string(pipeline_states.vad_probabilities(i)) + " ";
+    }
+    if(!pipeline_states_logs_){
+      pipeline_states_logs_.open("riva_asr_pipeline_states.log");
+    }
+    pipeline_states_logs_ << "VAD states: " << vad_log << std::endl;
+  } else {
+    bool is_final = result.is_final();
```
Comment (virajkarandikar) [inline riva/clients/asr/client_call.cc:42]: why is this moved to `else` ?

File: riva/clients/nmt/riva_nmt_t2t_client.cc [hunk 3]
```diff
@@ -200,6 +200,7 @@ main(int argc, char** argv)
   if (FLAGS_text != "") {
     nr_nmt::TranslateTextRequest request;
     nr_nmt::TranslateTextResponse response;
+    VLOG(1) << "Setting up t2t config.";
```
Comment (virajkarandikar) [inline riva/clients/nmt/riva_nmt_t2t_client.cc:203]: need this ?

File: riva/clients/asr/client_call.cc [hunk 4]
```diff
@@ -13,45 +13,63 @@ ClientCall::ClientCall(uint32_t corr_id, bool word_time_offsets)
...
-      latest_result_.final_transcripts[a] += result.alternatives(a).transcript();
-      latest_result_.final_scores[a] += result.alternatives(a).confidence();
+  if (result.has_pipeline_states()) {
+    auto pipeline_states = result.pipeline_states();
+    int prob_states_count = pipeline_states.vad_probabilities_size();
+    std::string vad_log = "";
+    for (int i = 0; i < prob_states_count; i++) {
+      vad_log += std::to_string(pipeline_states.vad_probabilities(i)) + " ";
+    }
+    if(!pipeline_states_logs_){
```
Comment (atomer-nvidia) [inline riva/clients/asr/client_call.cc]: Formatting: `if (!pipeline_states_logs_) {`

For future reference, you can make your changes, `git add` them, and run `git-clang-format`. It would fix all the formatting issues by itself. Then add the changed files again and commit.
https://github.com/llvm/llvm-project/blob/main/clang/tools/clang-format/git-clang-format

File: riva/clients/asr/client_call.cc (hunk 2 above)
Comment (sarane22) [reply riva/clients/asr/client_call.cc:42]: The VAD probabilties and transcripts are sent separately to reduce the AM processing time from VAD probs response latency.

File: WORKSPACE [hunk 5]
```diff
@@ -70,7 +70,7 @@ grpc_extra_deps()
 git_repository(
      name = "nvriva_common",
      remote = "https://github.com/nvidia-riva/common.git",
-     commit = "1c7da5aed4e4df3a296d2672379c5099a193aaae"
+     commit = "9b31412dc43a15740f5f55a97cbd8c3eb5b43d86"
```
Comment (atomer-nvidia) [inline WORKSPACE]: TODO: Need to changes this once common's commit gets merged.

File: WORKSPACE [hunk 6]
```diff
@@ -70,7 +70,7 @@ grpc_extra_deps()
 git_repository(
      name = "nvriva_common",
      remote = "https://github.com/nvidia-riva/common.git",
-     commit = "1c7da5aed4e4df3a296d2672379c5099a193aaae"
+     commit = "75afa27124dde29a5e892e6817d9841c6cbc49e7"
```
Comment (atomer-nvidia) [inline WORKSPACE]: This needs to change once common's change is committed?

File: riva/clients/asr/client_call.cc [hunk 7]
```diff
@@ -13,45 +13,63 @@ ClientCall::ClientCall(uint32_t corr_id, bool word_time_offsets)
...
-    latest_result_.final_transcripts.resize(num_alternatives);
-    latest_result_.final_scores.resize(num_alternatives);
-    latest_result_.final_time_stamps.resize(num_alternatives);
-    for (int a = 0; a < num_alternatives; ++a) {
-      // Append to transcript
-      latest_result_.final_transcripts[a] += result.alternatives(a).transcript();
-      latest_result_.final_scores[a] += result.alternatives(a).confidence();
+  if (result.has_pipeline_states()) {
+    auto pipeline_states = result.pipeline_states();
+    int prob_states_count = pipeline_states.vad_probabilities_size();
```
Comment (atomer-nvidia) [inline riva/clients/asr/client_call.cc]: Technically size() returns size_t datatype. If the vad_probabilities_size() hits int limits, it would roll over and return negative values. It's better to use size_t instead of int.

File: riva/clients/asr/client_call.cc [hunk 8]
```diff
@@ -13,45 +13,63 @@ ClientCall::ClientCall(uint32_t corr_id, bool word_time_offsets)
...
+    std::string vad_log = "";
+    for (int i = 0; i < prob_states_count; i++) {
```
Comment (atomer-nvidia) [inline riva/clients/asr/client_call.cc]: Use size_t instead to avoid compiler warnings.

File: riva/clients/asr/client_call.cc [hunk 9]
```diff
@@ -13,45 +13,63 @@ ClientCall::ClientCall(uint32_t corr_id, bool word_time_offsets)
...
+      vad_log += std::to_string(pipeline_states.vad_probabilities(i)) + " ";
+    }
+    if(!pipeline_states_logs_){
+      pipeline_states_logs_.open("riva_asr_pipeline_states.log");
```
Comment (atomer-nvidia) [inline riva/clients/asr/client_call.cc:38]: Who or when does it get cleaned up? Does this state needs persistence across the calls? Since the default mode for open is `out`, it will wipe the contents every time the client makes a call. Is that intended?

Comment (atomer-nvidia) [inline riva/clients/asr/client_call.cc:38]: Would it be better if we take output file from user instead of hardcoding it to riva_asr_pipeline_states.log? Or we would need to document that vad probabilities are dumped in xxx file.

File: WORKSPACE (hunk 6 above)
Comment (sarane22) [reply WORKSPACE]: yes

File: riva/clients/asr/client_call.cc (hunk 9 above)
Comment (sarane22) [reply riva/clients/asr/client_call.cc:38]: I am creating a file if the file pointer is not initialized. The client call object is shared across the all the streams, so file will be generated only once. Logs will get replace for next client run

Comment (sarane22) [reply riva/clients/asr/client_call.cc:38]: Idea was to dump the VAD states in a file instead of dumping them on the terminal.

> Would it be better if we take output file from user instead of hardcoding it to riva_asr_pipeline_states.log

This sounds good!!

Comment (sarane22) [reply riva/clients/asr/client_call.cc:38]: No, correct...... it's getting recreated by every thread and for every iteration.

Comment (atomer-nvidia) [reply riva/clients/asr/client_call.cc:38]: Lets keep it as is. Since this is part of additional custom_config, and is not hit in general code flow. This would be an explicit usecase and if someone using this config, they would know how/where to interpret/consume them.

File: riva/clients/nmt/riva_nmt_t2t_client.cc (hunk 3 above)
Comment (atomer-nvidia) [reply riva/clients/nmt/riva_nmt_t2t_client.cc:203]: I guess this was to fix a build failure. An unrelated build failure is occurring unless some superficial print is added to the nmt module. @sarane22 Is that correct?

Comment (sarane22) [reply riva/clients/nmt/riva_nmt_t2t_client.cc:203]: Yes correct.

**Review by atomer-nvidia — APPROVED** [review]
